File: dataforge-core/src/main/java/hep/dataforge/context/ContextLock.java
package hep.dataforge.context;

import hep.dataforge.exceptions.ContextLockException;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;

/**
 * Lock class for context
 */
public class ContextLock implements ContextAware {

    private final Context context;
    /**
     * A set of objects that lock this context
     */
    private final Set<Object> lockers = new HashSet<>();

    public ContextLock(Context context) {
        this.context = context;
    }

    public synchronized void lock(Object object) {
        this.lockers.add(object);
    }

    public synchronized void unlock(Object object) {
        this.lockers.remove(object);
    }

    public boolean isLocked() {
        return !lockers.isEmpty();
    }

    /**
     * Throws {@link ContextLockException} if context is locked
     */
    public void tryModify(){
        lockers.stream().findFirst().ifPresent(lock -> {
            throw new ContextLockException(lock);
        });
    }

    /**
     * Apply thread safe lockable object modification
     *
     * @param mod
     */
    public synchronized <T> T modify(Callable<T> mod) {
        tryModify();
        try {
            return context.getDispatcher().submit(mod).get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    public synchronized void modify(Runnable mod) {
        tryModify();
        try {
            context.getDispatcher().submit(mod).get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Context getContext() {
        return context;
    }
}

File: dataforge-core/src/main/kotlin/hep/dataforge/context/ContextLock.kt
package hep.dataforge.context

import hep.dataforge.exceptions.ContextLockException
import java.util.*
import java.util.concurrent.ExecutionException

/**
 * Lock class for context
 */
class ContextLock(override val context: Context) : ContextAware {
    /**
     * A set of objects that lock this context
     */
    private val lockers = HashSet<Any>()

    val isLocked: Boolean
        get() = !lockers.isEmpty()

    @Synchronized
    fun lock(`object`: Any) {
        this.lockers.add(`object`)
    }

    @Synchronized
    fun unlock(`object`: Any) {
        this.lockers.remove(`object`)
    }

    /**
     * Throws [ContextLockException] if context is locked
     */
    fun tryModify() {
        lockers.stream().findFirst().ifPresent { lock -> throw ContextLockException(lock) }
    }

    /**
     * Apply thread safe lockable object modification
     *
     * @param mod
     */
    @Synchronized
    fun <T> modify(mod: ()->T): T {
        tryModify()
        try {
            return context.dispatcher.submit(mod).get()
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        } catch (e: ExecutionException) {
            throw RuntimeException(e)
        }

    }

    @Synchronized
    fun modify(mod: () -> Unit) {
        tryModify()
        try {
            context.dispatcher.submit(mod).get()
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        } catch (e: ExecutionException) {
            throw RuntimeException(e)
        }

    }
}
